### PR TITLE
[FLINK-18986][kubernetes] Create ClusterClient only for attached deployments

### PR DESCRIPTION
**Pull request by zentol**
Only create a ClusterClient for attached deployments.

The client is not necessary for detached deployments, and the creation results in error messages if the REST API is not accessible from the outside.

I would like to add a test for this, but this requires mocking of several classes, and I only wanted to do that once I have approval of the change in general.

### CODE REVIEW

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java [hunk 1]
```diff
@@ -96,40 +97,51 @@ private int run(String[] args) throws FlinkException, CliArgsException {
...
+			boolean connectToExistingCluster = false;
+			try (final FlinkKubeClient kubeClient = KubeClientFactory.fromConfiguration(configuration)) {
+				connectToExistingCluster = clusterId != null && kubeClient.getRestService(clusterId).isPresent();
+			} catch (Exception e) {
+				LOG.info("Could not properly shutdown cluster client.", e);
+			}
 
 			// Retrieve or create a session cluster.
-			if (clusterId != null && kubeClient.getRestService(clusterId).isPresent()) {
-				clusterClient = kubernetesClusterDescriptor.retrieve(clusterId).getClusterClient();
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: Why is calling `getClusterClient` a problem? Which call does the `RestClusterClient` do where it cannot reach the cluster?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @zentol. Could you please explain how this change solves the described problem? Concretely, I would like to understand what the `RestClusterClient` is doing which causes a failed connection attempt to be logged.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java (hunk 1 above)
Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: It's not the RestClusterClient that is the issue, but the retrieval of the REST API address and port (https://github.com/apache/flink/blob/master/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java#L96).

Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: Alright, this makes sense. Thanks for clarification.

Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: Hmm ok, I see the problem. Maybe the problem is that we need to create a `ClusterClient` in order to retrieve the `clusterId` even though this is something which is known before the `ClusterClient` is being created.

Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: I guess my point is that one could already retrieve the `clusterId` from the `ClusterClientProvider` given the current implementations because the `RestClusterClient` is being initialized with `clusterId` which it simply returns.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java [hunk 2]
```diff
@@ -96,40 +97,51 @@ private int run(String[] args) throws FlinkException, CliArgsException {
...
 
-			try {
-				if (!detached) {
-					Tuple2<Boolean, Boolean> continueRepl = new Tuple2<>(true, false);
-					try (BufferedReader in = new BufferedReader(new InputStreamReader(System.in))) {
-						while (continueRepl.f0) {
-							continueRepl = repStep(in);
-						}
+			if (!detached) {
+				if (!connectToExistingCluster) {
```
Comment (zentol) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:121]:
```suggestion
				if (clusterId == null) {
```

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java (hunk 1 above)
Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: hmm....is it a hard requirement that a cluster-id must be specified when starting a cluster, or can it also be auto-generated by Kubernetes?

If it is not required, then we could adjust the PR such that the cluster-id is only queried if `clusterId` is null: https://github.com/apache/flink/pull/13190/files#r472915726

Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: Actually we can make this change in any case.

Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: Turns the clusterId is never null. `KubernetesClusterClientFactory#createClusterDescriptor` ensures that the cluster-id is set in the configuration, so we actually never need the client here.

Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: It's a bit annoying to rely on it as is since it requires methods to be called in a specific order though,

Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:106]: @tillrohrmann I've adjusted the code to explicitly set the cluster ID directly in the KubernetesSessionCli, and simplified the rest accordingly.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java [hunk 3]
```diff
@@ -96,40 +97,42 @@ private int run(String[] args) throws FlinkException, CliArgsException {
...
+					while (continueRepl.f0) {
+						continueRepl = repStep(in);
 					}
-					if (continueRepl.f1) {
+				} catch (Exception e) {
+					LOG.warn("Exception while running the interactive command line interface.", e);
+				}
+
+				if (continueRepl.f1) {
+					try {
```
Comment (zentol) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:130]: these changes are now purely about better exception messages, and I would exclude them from the final PR, or move them into a hotfix commit.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java [hunk 4]
```diff
@@ -88,6 +88,7 @@ public Configuration getEffectiveConfiguration(String[] args) throws CliArgsExce
 
 	private int run(String[] args) throws FlinkException, CliArgsException {
 		final Configuration configuration = getEffectiveConfiguration(args);
+		KubernetesClusterClientFactory.ensureClusterIdIsSet(configuration);
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:91]: Shouldn't we obtain the `clusterId` from the `KubernetesClusterDescriptor`?

Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:91]: That way we wouldn't have this clumsy call here.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java [hunk 5]
```diff
@@ -51,10 +51,7 @@ public boolean isCompatibleWith(Configuration configuration) {
 	@Override
 	public KubernetesClusterDescriptor createClusterDescriptor(Configuration configuration) {
 		checkNotNull(configuration);
-		if (!configuration.contains(KubernetesConfigOptions.CLUSTER_ID)) {
-			final String clusterId = generateClusterId();
-			configuration.setString(KubernetesConfigOptions.CLUSTER_ID, clusterId);
-		}
+		ensureClusterIdIsSet(configuration);
 		return new KubernetesClusterDescriptor(configuration, KubeClientFactory.fromConfiguration(configuration));
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java:55]: It is a bit unrelated. But setting a fixed cluster id for the `ClusterDescriptor` seems wrong. That way we can ever deploy a single cluster using a `ClusterDescriptor`. I believe that the `YarnClusterDescriptor` works differently.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating this PR @zentol. I had a comment concerning the changes.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java (hunk 5 above)
Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java:55]: Yes I noticed that too, it is at odds with the interface. You can connect to _any_ cluster, but only deploy a single one.
But this seems to be an issue of the `ClusterDescriptor` interface; there is no way for implementations to accept additional arguments for deploying clusters.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java (hunk 4 above)
Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:91]: The `(Kubernetes)ClusterDescriptor` provides no way of accessing the cluster-id.
There is `ClusterClientFactory#getClusterId`, but this one only works the way we'd like to after `ClusterClientFactory#createClusterDescriptor` was called (since it modifies the passed configuration).

I thought about changing `ClusterClientFactory#getClusterId` to also set the id if it is missing, to ensure that the behavior of the methods is identical regardless of method call order, but I cannot assess what repercussions this might have on users of the `ClusterClientFactory` interface.

Basically, I hate this behavior, and didn't want to rely on it.
I would argue that the user of the `ClusterClientFactory` should ensure that the cluster-id was set, but I didn't want to remove the auto-generation in `ClusterClientFactory#createClusterDescriptor` because god knows what it might break.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java (hunk 5 above)
Comment (zentol) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java:55]: The YarnClusterDescriptor actually works quite similarly; there things like custom names and application types are being defined in the constructor.
The `ClusterDescriptor` just doesn't handle this case very well, but admittedly we do not really have a use-case for deploying multiples clusters.

Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java:55]: I think the idea was that the `ClusterDescriptor` will generate a cluster id when you call any of the `deploy*` methods. I believe what is set in the constructor of the `YarnClusterDescriptor` is fine and does not contradict what I was saying.

If one really needs to configure the cluster id manually, then one could extend the `ClusterDescriptor` interface accordingly.

I think that we are adding technical debt by working around broken contracts which might be fine now but we should definitely pull this straight eventually.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:91]: Hmm, yeah one should either get the `clusterId` from the user or after a cluster is deployed via the `ClusterClient` or `ClusterClientProvider`.
